kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/throwable/ThrowableMatchersTest.kt

```diff
@@ -0,0 +1,158 @@
+      }
+      "shouldHaveCause" {
+        shouldThrowExactly<CompleteTestException> { throw CompleteTestException() }.shouldHaveCause()
+        shouldThrowExactly<CompleteTestException> { throw CompleteTestException() }.shouldHaveCause {
+          it shouldHaveMessage "file.txt not found"
+        }
+      }
+      "shouldNotHaveCause" {
+        shouldThrowExactly<TestException> { throw TestException() }.shouldNotHaveCause()
+        shouldThrowExactly<FileNotFoundException> { throw FileNotFoundException("this_file.txt not found") }.shouldNotHaveCause()
+      }
+      "shouldHaveCauseInstanceOf" {
+        shouldThrowExactly<CompleteTestException> { throw CompleteTestException() }.shouldHaveCauseInstanceOf<FileNotFoundException>()
+        shouldThrowExactly<CompleteTestException> { throw CompleteTestException() }.shouldHaveCauseInstanceOf<IOException>()
+      }
+      "shouldNotHaveCauseInstanceOf" {
+        shouldThrowExactly<CompleteTestException> { throw CompleteTestException() }.shouldNotHaveCauseInstanceOf<TestException>()
+      }
+      "shouldHaveCauseOfType" {
+        shouldThrowExactly<CompleteTestException> { throw CompleteTestException() }.shouldHaveCauseOfType<FileNotFoundException>()
+      }
+      "shouldNotHaveCauseOfType" {
+        shouldThrowExactly<CompleteTestException> { throw CompleteTestException() }.shouldNotHaveCauseOfType<IOException>()
+      }
+    }
+    "result" - {
+      "shouldHaveMessage" {
+        Result.failure<Any>(FileNotFoundException("this_file.txt not found")).exceptionOrNull()!! shouldHaveMessage "this_file.txt not found"
+        Result.failure<Any>(TestException()).exceptionOrNull()!! shouldHaveMessage "This is a test exception"
+        Result.failure<Any>(CompleteTestException()).exceptionOrNull()!! shouldHaveMessage "This is a complete test exception"
+      }
+      "shouldNotHaveMessage" {
+        Result.failure<Any>(FileNotFoundException("this_file.txt not found")).exceptionOrNull()!! shouldNotHaveMessage "random message"
+        Result.failure<Any>(TestException()).exceptionOrNull()!! shouldNotHaveMessage "This is a complete test exception"
+        Result.failure<Any>(CompleteTestException()).exceptionOrNull()!! shouldNotHaveMessage "This is a test exception"
+      }
+      "shouldHaveCause" {
+        Result.failure<Any>(CompleteTestException()).exceptionOrNull()!!.shouldHaveCause()
+        Result.failure<Any>(CompleteTestException()).exceptionOrNull()!!.shouldHaveCause {
+          it shouldHaveMessage "file.txt not found"
+        }
+      }
+      "shouldNotHaveCause" {
+        Result.failure<Any>(TestException()).exceptionOrNull()!!.shouldNotHaveCause()
+        Result.failure<Any>(FileNotFoundException("this_file.txt not found")).exceptionOrNull()!!.shouldNotHaveCause()
+      }
+      "shouldHaveCauseInstanceOf" {
+        Result.failure<Any>(CompleteTestException()).exceptionOrNull()!!.shouldHaveCauseInstanceOf<FileNotFoundException>()
```

Comment:
What's the difference between instanceOf and ofType?

Reply:
One check if is an instance of the provided class or it's parent (instanceOf), the other (typeOf) check if is exactly the same class. You can take the base of `shouldBeInstanceOf` and `shouldBeTypeOf` for any type.